In [1]:
import pandas as pd 

In [3]:
# upload all the files here 
rating = pd.read_csv(r"D:\\Asstha\\streamflix project\\ff5a102e4361911d42cef9e21cda23d0\\ratings.csv")
reviews = pd.read_csv(r"D:\\Asstha\\streamflix project\\ff5a102e4361911d42cef9e21cda23d0\\reviews.csv")
subscribers = pd.read_csv(r"D:\\Asstha\\streamflix project\\ff5a102e4361911d42cef9e21cda23d0\\subscribers.csv")
watch_history = pd.read_csv(r"D:\\Asstha\\streamflix project\\ff5a102e4361911d42cef9e21cda23d0\\watch_history.csv")
watchlist = pd.read_csv(r"D:\\Asstha\\streamflix project\\ff5a102e4361911d42cef9e21cda23d0\\watchlist.csv")

In [8]:
rating.head()

,rating_id,subscriber_id,title_id,rating,rating_date
0,1,SUB109184,TTL203909,4,2026-02-06
1,2,SUB105194,TTL200888,4,2025-03-21
2,3,SUB102974,TTL200036,3,2024-05-26
3,4,SUB105612,TTL206358,5,2026-05-22
4,5,SUB103070,TTL205227,4,2026-05-16


In [9]:
rating.shape

(130000, 5)

In [10]:
rating.dtypes

rating_id        int64
subscriber_id      str
title_id           str
rating           int64
rating_date        str
dtype: object

In [11]:
rating["rating_date"]= pd.to_datetime(rating["rating_date"])

In [13]:
rating.dtypes

rating_id                 int64
subscriber_id               str
title_id                    str
rating                    int64
rating_date      datetime64[us]
dtype: object

In [14]:
rating.isnull().sum()

rating_id        0
subscriber_id    0
title_id         0
rating           0
rating_date      0
dtype: int64

In [16]:
rating.duplicated().sum()

np.int64(0)

In [18]:
reviews.head()

,review_id,subscriber_id,title_id,review_text,sentiment,helpful_votes,review_date
0,1,SUB102409,TTL206425,Stellar cast and a satisfying ending.,Positive,9,2023-11-04
1,2,SUB111270,TTL203987,Predictable plot and flat characters.,Negative,4,2026-04-25
2,3,SUB107848,TTL203235,"Started strong, lost steam in the middle.",Neutral,2,2024-07-31
3,4,SUB114501,TTL200930,Beautifully shot and emotionally resonant.,Positive,14,2022-12-13
4,5,SUB113370,TTL207141,A rare gem. Highly recommend.,Positive,6,2025-10-09


In [20]:
reviews.shape

(110000, 7)

In [21]:
reviews.dtypes

review_id        int64
subscriber_id      str
title_id           str
review_text        str
sentiment          str
helpful_votes    int64
review_date        str
dtype: object

In [22]:
reviews["review_date"] = pd.to_datetime(reviews["review_date"])

In [23]:
reviews.dtypes

review_id                 int64
subscriber_id               str
title_id                    str
review_text                 str
sentiment                   str
helpful_votes             int64
review_date      datetime64[us]
dtype: object

In [24]:
reviews.head()

,review_id,subscriber_id,title_id,review_text,sentiment,helpful_votes,review_date
0,1,SUB102409,TTL206425,Stellar cast and a satisfying ending.,Positive,9,2023-11-04
1,2,SUB111270,TTL203987,Predictable plot and flat characters.,Negative,4,2026-04-25
2,3,SUB107848,TTL203235,"Started strong, lost steam in the middle.",Neutral,2,2024-07-31
3,4,SUB114501,TTL200930,Beautifully shot and emotionally resonant.,Positive,14,2022-12-13
4,5,SUB113370,TTL207141,A rare gem. Highly recommend.,Positive,6,2025-10-09


In [25]:
reviews.isnull().sum()

review_id        0
subscriber_id    0
title_id         0
review_text      0
sentiment        0
helpful_votes    0
review_date      0
dtype: int64

In [26]:
reviews.isnull().mean()*100

review_id        0.0
subscriber_id    0.0
title_id         0.0
review_text      0.0
sentiment        0.0
helpful_votes    0.0
review_date      0.0
dtype: float64

In [27]:
reviews.duplicated().sum()

np.int64(0)

In [28]:
subscribers.head()

,subscriber_id,signup_date,country,region,age,gender,plan_type,monthly_price_usd,household_size,primary_device,payment_method,tenure_months,is_active,churn_date
0,SUB100000,2023-01-09,France,Europe,22,Female,Basic with Ads,6.99,1,Mobile,Credit Card,41,True,NaN
1,SUB100001,2020-02-05,United Kingdom,Europe,48,Male,Basic with Ads,6.99,1,Smart TV,Credit Card,76,False,2022-12-08
2,SUB100002,2022-04-15,Italy,Europe,35,Male,Standard,15.49,2,Streaming Stick,Gift Card,50,True,NaN
3,SUB100003,2024-10-10,Spain,Europe,49,Female,Basic with Ads,6.99,1,Laptop,Credit Card,19,True,NaN
4,SUB100004,2020-04-12,United States,North America,36,Female,Basic with Ads,6.99,1,Mobile,Credit Card,74,False,2021-12-15


In [29]:
subscribers.shape

(15000, 14)

In [30]:
subscribers.dtypes

subscriber_id            str
signup_date              str
country                  str
region                   str
age                    int64
gender                   str
plan_type                str
monthly_price_usd    float64
household_size         int64
primary_device           str
payment_method           str
tenure_months          int64
is_active               bool
churn_date               str
dtype: object

In [31]:
subscribers["signup_date"] = pd.to_datetime(subscribers["signup_date"])
subscribers["churn_date"] = pd.to_datetime(subscribers["churn_date"])

In [32]:
subscribers.dtypes

subscriber_id                   str
signup_date          datetime64[us]
country                         str
region                          str
age                           int64
gender                          str
plan_type                       str
monthly_price_usd           float64
household_size                int64
primary_device                  str
payment_method                  str
tenure_months                 int64
is_active                      bool
churn_date           datetime64[us]
dtype: object

In [33]:
subscribers.isnull().sum()

subscriber_id            0
signup_date              0
country                  0
region                   0
age                      0
gender                   0
plan_type                0
monthly_price_usd        0
household_size           0
primary_device           0
payment_method           0
tenure_months            0
is_active                0
churn_date           11199
dtype: int64

In [34]:
subscribers.duplicated().sum()

np.int64(0)

In [35]:
subscribers.isnull().mean()*100

subscriber_id         0.00
signup_date           0.00
country               0.00
region                0.00
age                   0.00
gender                0.00
plan_type             0.00
monthly_price_usd     0.00
household_size        0.00
primary_device        0.00
payment_method        0.00
tenure_months         0.00
is_active             0.00
churn_date           74.66
dtype: float64

In [40]:
subscribers[subscribers["churn_date"].isnull()]["is_active"].value_counts()

is_active
True    11199
Name: count, dtype: int64

## churn_date contains 11,199 missing values (74.66%). All subscribers with a missing churn_date have is_active = True, indicating that these subscribers have not churned yet. Therefore, these missing values are considered expected/valid rather than a data-quality issue.


In [41]:
subscribers["subscriber_id"].duplicated().sum()

np.int64(0)

In [44]:
## as per the question asked we have to check that churn date should not be less than the sign up date 
subscribers[subscribers["churn_date"] < subscribers["signup_date"]]

,subscriber_id,signup_date,country,region,age,gender,plan_type,monthly_price_usd,household_size,primary_device,payment_method,tenure_months,is_active,churn_date


In [45]:
subscribers["plan_type"].value_counts()

plan_type
Standard          6100
Basic with Ads    4458
Premium           4442
Name: count, dtype: int64

In [46]:
subscribers["is_active"].value_counts()

is_active
True     11199
False     3801
Name: count, dtype: int64

SUBSCRIBERS – DATA QUALITY FINDINGS

• No duplicate rows found.
• No duplicate subscriber_id values found.
• No missing values except churn_date.
• churn_date has 11,199 missing values (74.66%).
• All subscribers with missing churn_date are active, so these nulls are logically valid.
• signup_date and churn_date were converted to datetime.
• No records were found where churn_date was earlier than signup_date.
• plan_type contains only the expected three plans:
  Standard, Basic with Ads, and Premium.
• is_active contains valid values.
